### PR TITLE
rgw/lua: support packages via luarocks

### PR DESCRIPTION
**Pull request by yuvalif**
Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

## Testing
To add a module (e.g. json, which is part of the [luajson](https://github.com/harningt/luajson) package) to the allowlist use the following command:
```
radosgw-admin script-package add --package luajson
radosgw-admin script-package add --package luasocket --allow-compilation
```
To add a package that contains C source code that needs to be compiled, use the `--allow-compilation` flag.
Packages will be uploaded to the RGW when it restarts, so, in the case of vstart cluster, it will have to be stopped and then started without the `-n` flag.
Following shows an example a script that does "access log" on RGW by sending json objects over a unix socket (similar to the ops log):
```lua
if Request.RGWOp == "get_obj" then
  local json = require("json")
  local socket = require("socket")
  local unix = require("socket.unix")
  local s = assert(unix())

  E = {}

  msg = {bucket = (Request.Bucket or (Request.CopyFrom or E).Bucket).Name,
  time = Request.Time,
  operation = Request.RGWOp,
  http_status = Request.Response.HTTPStatusCode,
  error_code = Request.Response.HTTPStatus,
  object_size = Request.Object.Size,
  trans_id = Request.TransactionId}

  assert(s:connect("/tmp/socket"))
  assert(s:send(json.encode(msg).."\n"))
  assert(s:close())
end
```

## TODO
- [x] support module removal
- [x] add unix socket + json example for access log
- [x] add documentation
- [x] support configurable luarocks package directory configurable
- [x] verify behavior when disabled via compilation flag
- [x] change in all places "modules"->"packages"
- [x] create per RGW subdirectory
- (optional) make [luasyslog](https://github.com/ntd/luasyslog/) example work
- (optional) allow module reload via radosgw-admin command
- (optional) yield while modules are re/loaded

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_lua.cc
Comment (tchaikov) [inline src/rgw/rgw_lua.cc:232]: might want to use `find_package(Lua)` for its version number.

File: debian/control
Comment (tchaikov) [inline debian/control:63]: please use `find_package(Lua)` for checking the existence of Lua libraries. could you please double check if the runtime dependencies related to Lua are added to `radosgw` by `${shlibs:Depends}` ?

File: CMakeLists.txt
Comment (tchaikov) [inline CMakeLists.txt:570]: i don't understand why we need Boost.filesystem? could you shed some light on this?

File: src/rgw/rgw_admin.cc
Comment (tchaikov) [inline src/rgw/rgw_admin.cc:9410]: s/adding/listing/

why not just bail out at seeing unknown options?

File: src/rgw/rgw_lua_utils.cc
Comment (tchaikov) [inline src/rgw/rgw_lua_utils.cc:48]: might want to replace magic numbers with version numbers detected using `find_package()`.

File: src/rgw/rgw_lua.h
Comment (tchaikov) [inline src/rgw/rgw_lua.h:42]: should this be located somewhere under `/var/lib` instead of hiding it in a dot directory ?

Comment (yuvalif) [reply src/rgw/rgw_lua.h:42]: but how would that work in case of a vstart cluster?
in theory, this is just a cache, as the needed lua rocks are installed on every restart of the RGW.
maybe we should put that under `/tmp`? or based on an env var?

File: CMakeLists.txt
Comment (yuvalif) [reply CMakeLists.txt:570]: this is required by `boost::process` - I use it to execute `luarocks`

File: debian/control
Comment (yuvalif) [reply debian/control:63]: in the code I don't need lua (or lua-dev) since we have a lua submodule which provides header and lib that we need for compilation.
however, the luarocks binary does need the lua header and lib in the cases that it compiles C code (some lua rocks require compilation, and I added a special flag for that).

I guess that there are several options here:
1. keep the submodule, and install lua-dev for luarocks to work (what I currently did)
2. keep the submodule, and install header and lib from the submodule (currently they are not installed and just used internally) - not 100% sure on how to do that, especially in the vstart case
3. remove the submodule (need to make sure that we don't break cls_lua), and use `find_package(Lua)` instead internally, and install  lua-dev for luarocks

File: src/rgw/rgw_admin.cc
Comment (yuvalif) [reply src/rgw/rgw_admin.cc:9410]: > s/adding/listing/
> 
+1

> why not just bail out at seeing unknown options?

it will require ifdefing in lots of places

**Comment by yuvalif** [comment]
@batrick regarding https://github.com/ceph/ceph/pull/37924#discussion_r516400082, do you know if there is a good reason to keep lua as a submodule?

File: src/rgw/rgw_lua.h
Comment (tchaikov) [reply src/rgw/rgw_lua.h:42]: > but how would that work in case of a vstart cluster?

we have the same issue when it comes to asok, log files and more. we can address this problem using a dedicated option. i don't think `/tmp` is the right place to cache lua modules. according to [FHS](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html), `/var/lib` is the right place.

File: debian/control
Comment (batrick) [reply debian/control:63]: I think (2) is a non-starter. We don't want to fake installation or direct compiled modules to use our Lua source. I don't have a strong opinion on (1) or (3). I would defer to @tchaikov 's preference. Or maybe he has an alternative.

Comment (yuvalif) [reply debian/control:63]: looking at the list of changes in the lua submodule: https://github.com/LuaDist/lua/compare/lua-5.1...ceph:lua-5.3-ceph
it looks like there was one actual change to the code in 2013 (fixing a memory leak). the rest of the changes are around cmake and the build environment.
latest major change is merge from the lua5.3 branch in 2015.
if we can get rid of this submodule, we should probably go with option (3)

Comment (yuvalif) [reply debian/control:63]: trying it out here: https://github.com/ceph/ceph/pull/37940

File: src/rgw/rgw_lua.h
Comment (yuvalif) [reply src/rgw/rgw_lua.h:42]: ok. according to this: https://github.com/ceph/ceph/blob/master/src/vstart.sh#L1526
I will have to add that as a parameter to the rgw executable

Comment (yuvalif) [reply src/rgw/rgw_lua.h:42]: > > but how would that work in case of a vstart cluster?
> 
> we have the same issue when it comes to asok, log files and more. we can address this problem using a dedicated option. i don't think `/tmp` is the right place to cache lua modules. according to [FHS](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html), `/var/lib` is the right place.

one of the intents in using only a local directory was to be separated from the environment of the host
- rocks installed on the host cannot be used by rgw if not in the allowlist
- rocks installed via the allowlist are not added to the host (this is also for the case that rgw is not allowed to change system directories)
- BTW, the official locations are:
 - `/usr/share/lua/` for the lua code of the rocks
- `/usr/lib/lua/` for .so files of the rocks
- `/usr/lib/luarocks/` for metadata

**Comment by batrick** [comment]
> @batrick regarding [#37924 (comment)](https://github.com/ceph/ceph/pull/37924#discussion_r516400082), do you know if there is a good reason to keep lua as a submodule?

Statically linking Lua can make a lot of sense for most project as it's small and gives us control over the exact version used. It also lets us skirt around the issue of e.g. Lua 5.4 not present on CentOS or Debian.

However, that may complicate luarocks installation. We would probably need to add a luarocks submodule too. That may be worth doing so we can have more control over it.

File: src/rgw/rgw_lua.cc
Comment (yuvalif) [reply src/rgw/rgw_lua.cc:232]: removing the version. since now `FindLua()` may get a version newer than 5.3

File: src/rgw/rgw_lua.h
Comment (yuvalif) [reply src/rgw/rgw_lua.h]: make the directory configurable. by default it is set to `/tmp/luarocks` so it will be separated from any lua rocks installed on the host. but could be set to another location

**Comment by yuvalif** [comment]
luarocks does not exists for centos8 (it was part of epel repo in centos7). not sure how to proceed here :-(
should I create a submodule?

**Comment by yuvalif** [comment]
> luarocks does not exists for centos8 (it was part of epel repo in centos7). opened a [ticket for epel](https://pagure.io/epel/issue/107).
> 
> there is no actual dependency between our code and luarocks (we just invoke the binary from our code), so a submodule is not really needed.
> for now, I can disable for rhel8 and centos8 :-( just so that the builds won't fail.
> 
> @tchaikov is it possible to create our own luarocks rpm package and use it for centos8?

this is fixed now thanks to @kalebskeithley !
see: https://github.com/ceph/ceph/pull/37924/files#diff-1335d27b188d672bf9a9b572f40df7c4699b8d9998fb16ba284f4acb4e77077c

File: install-deps.sh
Comment (tchaikov) [inline install-deps.sh:341]: i cannot find any explanation regarding this change in the commit message. could you elaborated a little bit?

File: install-deps.sh
Comment (tchaikov) [inline install-deps.sh:327]: could you please move this change into another commit and keep the indent?

Comment (yuvalif) [reply install-deps.sh]: done

**Comment by smithfarm** [comment]
The immediate problem I see (from a SUSE perspective) is that `lua53-luarocks` is not available in any official SLE module (only on the package hub). So this PR will need something like this in the part that constructs the cmake command:

```
%if 0%{?suse_version} && ! 0%{?is_opensuse}
    -DWITH_RADOSGW_LUA_MODULES=OFF
%endif
```

**Comment by yuvalif** [comment]
> The immediate problem I see (from a SUSE perspective) is that `lua53-luarocks` is not available in any official SLE module (only on the package hub). So this PR will need something like this in the part that constructs the cmake command:
> 
> ```
> %if 0%{?suse_version} && ! 0%{?is_opensuse}
>     -DWITH_RADOSGW_LUA_MODULES=OFF
> %endif
> ```

fixed. now it is enabled only in opensuse

**Comment by yuvalif** [comment]
> > luarocks does not exists for centos8 (it was part of epel repo in centos7). opened a [ticket for epel](https://pagure.io/epel/issue/107).
> > there is no actual dependency between our code and luarocks (we just invoke the binary from our code), so a submodule is not really needed.
> > for now, I can disable for rhel8 and centos8 :-( just so that the builds won't fail.
> > @tchaikov is it possible to create our own luarocks rpm package and use it for centos8?
> 
> this is fixed now thanks to @kalebskeithley !
> see: https://github.com/ceph/ceph/pull/37924/files#diff-1335d27b188d672bf9a9b572f40df7c4699b8d9998fb16ba284f4acb4e77077c

all builds are passing on latest in shaman: https://shaman.ceph.com/builds/ceph/yuval-wip-lua-packages/dbec100784a4b7178f8381019933b0c61a768cd3/

luarocks now added to lab-extras repo

File: ceph.spec.in
Comment (smithfarm) [inline ceph.spec.in:95]: This conditional makes my brain hurt. It seems like it's trying to say this:

```
%if 0%{?suse_version} && ! 0%{?is_opensuse}
# SLE does not support luarocks
%bcond_with lua_modules
%else
%global luarocks_package luarocks
%bcond_without lua_modules
%endif
```

or this:

```
%if ! 0%{?suse_version} || 0%{?is_opensuse}
%global luarocks_package luarocks
%bcond_without lua_modules
%else
# SLE does not support luarocks
%bcond_with lua_modules
%endif
```

Comment (yuvalif) [reply ceph.spec.in:95]: i mixed the package name logic in the same condition. but taking it away this should be:
```
 %if 0%{?suse_version} && 0%{?is_opensuse}
￼%bcond_without lua_modules
￼%else
￼# SLE does not support luarocks
￼%bcond_with lua_modules
￼%endif
```
it also puzzles me...  but  setting: `￼%bcond_without lua_modules` means that if you check `%if 0%{with lua_modules}` it would return `true`

Comment (smithfarm) [reply ceph.spec.in:95]: @yuvalif I am suggesting to replace the complicated nested conditional (with two `%else` clauses) with a single, easy-to-understand conditional (with just one `%else`) clause. The conditional you wrote would not replace the entire conditional.

And yes, the meaning of `%bcond_with` and `%bcond_without` is clear, though it certainly can be puzzling when one is faced with it for the first time.

Comment (smithfarm) [reply ceph.spec.in:95]: In other words, I'm suggesting to replace this nested conditional:

```
%if 0%{?suse_version}
%if 0%{?is_opensuse}
%global luarocks_package lua53-luarocks
%bcond_without lua_modules
%else
# SLE does not support luarocks
%bcond_with lua_modules
%endif
%else
%global luarocks_package luarocks
%bcond_without lua_modules
%endif
```

with one of the two simple variants I suggested - both of which are, I believe, logically equivalent to the nested conditional.

Comment (yuvalif) [reply ceph.spec.in:95]: @smithfarm how about the following:

```
%if 0%{?suse_version} && !0%{?is_opensuse}
# SLE does not support luarocks
%bcond_with lua_modules
%elif 0%{?suse_version} && 0%{?is_opensuse}
%global luarocks_package lua53-luarocks
%bcond_without lua_modules
%else
%global luarocks_package luarocks
%bcond_without lua_modules
%endif
```

Comment (smithfarm) [reply ceph.spec.in:95]: :shrug: Even that would be an improvement :-)

**Comment by yuvalif** [comment]
teuthology results: http://pulpito.front.sepia.ceph.com/yuvalif-2020-11-12_12:40:55-rgw-yuval-wip-lua-packages-distro-basic-smithi/
9 failures:
- barbican failure (expected)
- 2 `test_get_object_with_x_object_manifest` (expected)
- rest of the failures are in `radosgw_admin_rest` with:
```
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 90, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 69, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_yuval-wip-lua-packages/qa/tasks/radosgw_admin_rest.py", line 470, in task
    assert out['Code'] == 'NoSuchBucket'
```

File: ceph.spec.in
Comment (yuvalif) [reply ceph.spec.in:93]: "elif" was just recently implemented in rpm, and does not work on opensuse.
have to use nested if.

**Comment by yuvalif** [comment]
[teuthology](http://pulpito.front.sepia.ceph.com/yuvalif-2020-11-16_17:26:30-rgw-wip-yuval-lua-packages-distro-basic-smithi/) is all green except 2 `test_get_object_with_x_object_manifest` failures and one `PG_AVAILABILITY` error.

**Comment by sebastian-philipp** [comment]
Hm. we already have python based extension to ceph. Is adding yet another language the right thing to do? I once also did a lua integration into a c++ app. My experience was that lua is rather limited. We we ok with adding LUA, despite we already decided to use Python to extend Ceph?

**Comment by yuvalif** [comment]
> Hm. we already have python based extension to ceph. Is adding yet another language the right thing to do? I once also did a lua integration into a c++ app. My experience was that lua is rather limited. We we ok with adding LUA, despite we already decided to use Python to extend Ceph?

not sure its is the same usecase (don't think we have python scripting in RGW). 
Can you please provide more info on the usecases for these python extensions?
Also, would be interesting to know what limitations lua has in such integration?

**Comment by batrick** [comment]
> Hm. we already have python based extension to ceph. Is adding yet another language the right thing to do? I once also did a lua integration into a c++ app. My experience was that lua is rather limited. We we ok with adding LUA, despite we already decided to use Python to extend Ceph?

There's plenty of discussion on the internet about Lua vs. python so I'm not going to go into that in depth other than to say Lua is more appropriate as an embedded scripting language than Python. We do already use Lua in some contexts in Ceph: the OSD object classes and Mantle in the MDS.

File: doc/radosgw/lua-scripting.rst
Comment (batrick) [inline doc/radosgw/lua-scripting.rst:13]: How strong do you want this sandbox to be? With the `os` and `debug` packages fully available, any sandbox you create can be trivially broken out of.

File: src/common/options.cc
Comment (batrick) [inline src/common/options.cc:7226]:
```suggestion
   Option("rgw_luarocks_location", Option::TYPE_STR, Option::LEVEL_ADVANCED)
   .set_flag(Option::FLAG_STARTUP)
```

File: src/common/config.cc
Comment (batrick) [inline src/common/config.cc:716]: This argument needs documented. I would think this would go in the rgw code too as it's not really common to all daemons.

File: src/rgw/rgw_lua_utils.cc
Comment (batrick) [inline src/rgw/rgw_lua_utils.cc:61]:
```suggestion
  lua_getglobal(L, "package");
  if (!lua_istable(L, -1)) {
    return;
  }
  pushstring(L, path);
  lua_setfield(L, -2, "path"); 
  
  const auto cpath = install_dir+"/lib/lua/"+CEPH_LUA_VERSION+"/?.so";
  pushstring(L, cpath);
  lua_setfield(L, -2, "cpath"); 
```

Comment (batrick) [reply src/rgw/rgw_lua_utils.cc:61]: This also overrides the "default" search paths. Do you want that?

Comment (batrick) [reply src/rgw/rgw_lua_utils.cc:61]: Also: if the luarocks installation has other packages installed, your allowlist is circumvented. Should the rgw path always be some transient sandbox? What if it's the system directory/default?

**Review by batrick — CHANGES_REQUESTED** [review]

File: doc/radosgw/lua-scripting.rst
Comment (yuvalif) [reply doc/radosgw/lua-scripting.rst:13]: my intent is not to sandbox the users. since you need radosgw-admin privileges in order to push scripts, i can assume that that the author of the script will not write malicious scripts.
my intent is to allow organizations to control which libraries are safe to use.
I assume that the standard lua libraries are safe, so I allow them by default.

File: src/rgw/rgw_lua_utils.cc
Comment (yuvalif) [reply src/rgw/rgw_lua_utils.cc:61]: > This also overrides the "default" search paths. Do you want that?

yes, i want the rgw lua search path to be completely detached from whatever is configured and installed on the machine 

> Also: if the luarocks installation has other packages installed, your allowlist is circumvented. 

why? i don't search in the regular path

> Should the rgw path always be some transient sandbox? What if it's the system directory/default?

yes, by default it is set to `/tmp/luarocks`. note that i delete it before i install new packages to it (to more easily handle package deletions). if this directory is configured to the systems luarocks directory, i will mess up the packages on the machine. but this will just be misconfiguration

File: src/common/config.cc
Comment (yuvalif) [reply src/common/config.cc:716]: ok. will move it elsewhere.

File: src/rgw/rgw_lua_utils.cc
Comment (batrick) [reply src/rgw/rgw_lua_utils.cc:61]: > > This also overrides the "default" search paths. Do you want that?
> 
> yes, i want the rgw lua search path to be completely detached from whatever is configured and installed on the machine

K

> > Also: if the luarocks installation has other packages installed, your allowlist is circumvented.
> 
> why? i don't search in the regular path

Ok

> > Should the rgw path always be some transient sandbox? What if it's the system directory/default?
> 
> yes, by default it is set to `/tmp/luarocks`. note that i delete it before i install new packages to it (to more easily handle package deletions). if this directory is configured to the systems luarocks directory, i will mess up the packages on the machine. but this will just be misconfiguration

Perhaps you should catch that misconfiguration or at least document it.

Comment (yuvalif) [reply src/rgw/rgw_lua_utils.cc:61]: > Perhaps you should catch that misconfiguration or at least document it.

will document as part of this parameter's documentation.

File: doc/radosgw/lua-scripting.rst
Comment (batrick) [reply doc/radosgw/lua-scripting.rst:13]: > my intent is not to sandbox the users. since you need radosgw-admin privileges in order to push scripts, i can assume that that the author of the script will not write malicious scripts.

The sandbox provided should be documented, at least its philosophy.

> my intent is to allow organizations to control which libraries are safe to use.
> I assume that the standard lua libraries are safe, so I allow them by default.

Lua libraries are not really safe by default. It's expected that an embedding application trims down the libraries to what they need. I would filter out most of the `os` library. all of the `debug` library, `io.popen`, `load`, `loadfile`, `dofile`. Someone very determined could probably break out of the sandbox but there's no reason to make it easy. Keep in mind that someone could be an attacker that gained control of the script somehow.

Also, rather than filtering the libraries, I would create an allowlist of which functions to export and remove the rest.

Comment (yuvalif) [reply doc/radosgw/lua-scripting.rst:13]: > Lua libraries are not really safe by default. It's expected that an embedding application trims down the libraries to what they need. I would filter out most of the `os` library. all of the `debug` library, `io.popen`, `load`, `loadfile`, `dofile`. Someone very determined could probably break out of the sandbox but there's no reason to make it easy. Keep in mind that someone could be an attacker that gained control of the script somehow.
>

ok. this makes sense, would remove these standard calls:
* `os.exit()`
* `os.execute()`
* `io.popen()`
* `load()`
* `loadfile()`
* `loadstring()`
* `dofile()`
* the debug library

though I'm not 100% sure about `os.execuite()` and `io.popen()` as they may allow for interesting usecases 

> Also, rather than filtering the libraries, I would create an allowlist of which functions to export and remove the rest.

agree that this may be needed, but at this point I would say that whole or nothing approach towards the packages should be sufficient

File: src/common/config.cc
Comment (yuvalif) [reply src/common/config.cc]: removed from here

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins render docs

**Comment by yuvalif** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]
Lua bits are okay with me. I haven't tested the code.

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
> Lua bits are okay with me. I haven't tested the code.

thanks @batrick !

@cbodley @mattbenjamin @theanalyst need an approval from RGW team as well

File: src/rgw/rgw_admin.cc
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc:9377]: "in not permitted" should be "is not permitted' or just "not permitted"

File: src/rgw/rgw_lua.cc
Comment (mattbenjamin) [inline src/rgw/rgw_lua.cc:168]: what does lack of yield imply for this initial version?

File: src/rgw/rgw_lua.cc
Comment (mattbenjamin) [inline src/rgw/rgw_lua.cc:231]: generalizing for all of (add, remove, list);  packages appear to be known via omap, but are present in a local path;  can you explain how this works when we have a group of rgws on different hosts?

File: src/rgw/rgw_process.cc
Comment (mattbenjamin) [inline src/rgw/rgw_process.cc:228]: it would be nice if we could cache this somehow...either for the lifetime of radosgw, or the cached config option mechanism that has been around for a while

**Review by mattbenjamin — COMMENTED** [review]
this is brilliant;  a couple of nits, and one I think actionable request--cache script location called for each request

File: src/rgw/rgw_lua.cc
Comment (yuvalif) [reply src/rgw/rgw_lua.cc:168]: lack of yield will only be an issue if we allow package reload.
currently, we support only package load on startup, so it does not matter ( will change comment)

File: src/rgw/rgw_lua.cc
Comment (yuvalif) [reply src/rgw/rgw_lua.cc:231]: each RGW install the packages on its own host, later on to be used by the lua VM.

now I see that we have a problem with multiple RGWs running on the same host. will have to create separate directory for each RGW so they won't step on one another.

Comment (cbodley) [reply src/rgw/rgw_lua.cc:230]: you can use radosgw's `EntityName` as a stable/unique filesystem path. `conf->name.to_str()` will give you something like `client.rgw.0`

File: src/rgw/rgw_process.cc
Comment (yuvalif) [reply src/rgw/rgw_process.cc]: @mattbenjamin caching as a data member to `RGWRadosStore`.  here: https://github.com/ceph/ceph/pull/37924/commits/7ff10c504a7e015967873aa9dddbda84f03fe786#diff-9dec70d3aeed008aebebd862985b5d53466e1b55bdd6220b984ab099a9a29354R414
does it make sense?

**Review by dang — COMMENTED** [review]
Zipper bits look good.

**Review by dang — APPROVED** [review]

**Comment by ktdreyer** [comment]
@kalebskeithley where did you upload the luarocks RPM for Ceph?

Let's track luarocks in epel8 at https://bugzilla.redhat.com/show_bug.cgi?id=1760698

**Pull request merged**
